### DISCUSSION TITLE
Dataset Pre-Download Check

### DISCUSSION THREAD

**Discussion by FiYir** [unrealities/MTGA_Draft_17Lands#38] [Ideas]
17Lands doesn't currently make it easy to see if a dataset has been recently updated. Their FAQ says they generate daily stats starting at 00 UTC, with data becoming available sometime afterward. The app claims data is ready by 3:00 UTC, but that's sometimes optimistic. It's frustrating to spend minutes downloading a dataset only to find out it hasn't been updated.

A trick I use is checking the `# Games` column in the `All Decks` row on the [deck_color_data](https://www.17lands.com/deck_color_data) page. If the number has increased since my last check, I know the data's been updated. I think adding this check to the app would be helpful.

When downloading a dataset, we could include a `game_count` field in the file. The app would compare this `game_count` with the count from 17Lands endpoint and display a "Dataset is up-to-date" message if they're the same. It would only do this if the `Start Date` matches the stored dataset and the `End Date` is equal to or later than the stored dataset.

Users will likely abuse this by repeatedly starting downloads to check for new data. To manage this, we could impose a 60-second limit on how often this check can be done. The app would show a "Rate limit reached. Please wait {timeout} seconds and try again, or continue with the download." This way, we limit the checks without restricting access to downloads.

@unrealities What do you think of this solution?

**Comment by unrealities** [comment]
This is good. `game_count` is an excellent field to check against.

I'm curious about anything else we could do to ease hammering 17lands. Could we create an endpoint that caches the game_count? Then, whenever anyone downloads set data, we can update our cache.

They could create a simple endpoint that shows when the last time each set was updated or something similar.

**Reply by FiYir** [reply]
I had asked them about adding the update timestamp to the page and the endpoint, but they didn't seem interested.

![image](https://github.com/user-attachments/assets/9e64f7bf-830f-4a65-879b-0ca472e90a86)

**Comment by FiYir** [comment]
This is what I have so far.

![image](https://github.com/user-attachments/assets/9b6a6448-95cb-41b3-a040-2e3d76378ba2)
I modified the original download popup so that it includes the current time UTC.

![image](https://github.com/user-attachments/assets/f500ecb6-5dd6-4b86-b69b-0ae312c42b7a)
This is what you will see if the endpoint game count matches your stored dataset.

![image](https://github.com/user-attachments/assets/f69232e0-62cf-4515-8e0c-f9cfda0962eb)
This is what you will see if the endpoint game count is 0.

![image](https://github.com/user-attachments/assets/1dc3de06-1ffc-4391-a800-b6bda969dca1)
This is what you will see if you try to start another download within 60 seconds. 

@unrealities Any feedback?

**Reply by unrealities** [reply]
This looks great. The only possible upgrade for the first one would be to only show the alert if they don't already have the previous day's data. After the first couple of days of a set, I will only fetch data maybe twice a week and don't care too much if I'm not getting today's new data.

**Reply by FiYir** [reply]
Which of the four alerts are you referring to? The first one with the current time?

**Comment by unrealities** [comment]
Yes, sorry.  I think we should go with what you have.  It's more about the
situation where a set has been out for at least a few days and someone
hasn't downloaded it yet, probably doesn't require any alert, just let them
download. But having an alert before fetching a third party API is probably
good in general.

**Reply by FiYir** [reply]
Yeah, I think it's a good idea to keep the first alert, but I'm not sure about the message. I'm wondering if the part about the data availability should be moved to the "Your dataset is already up-to-date" alert, leaving the first message as "Are you sure you want to download....".

Maybe something like this?

![image](https://github.com/user-attachments/assets/a4b1beff-12d8-4885-b32a-2375ee750bbf)

![image](https://github.com/user-attachments/assets/3fe1ea78-2d7b-42be-af06-eda6fb1aec97)

Should we consider removing that part of the message entirely, given that we don't know when the new data will be available?

Also, should we replace "Add Sets" with "Download Dataset" everywhere it's used to ensure consistent terminology?

**Reply by unrealities** [reply]
I like the alerts you have here. And yes, I never liked "Add Sets," let's go with "Download Dataset."

**Comment by unrealities** [comment]
Released in v3.25
